File: isegoria/src/main/java/com/eulersbridge/isegoria/poll/PollOptionAdapter.java
package com.eulersbridge.isegoria.poll;

import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.common.LoadingAdapter;
import com.eulersbridge.isegoria.models.PollOption;

public class PollOptionAdapter extends LoadingAdapter<PollOption, PollOptionViewHolder> implements PollOptionViewHolder.ClickListener {

    public interface PollOptionVoteListener {
        /**
         * Called when any view of a poll option is clicked, but not when poll voting is disabled
         */
        void onPollOptionClick(int position);
    }

    private boolean pollVotingEnabled = true;

    final private @NonNull PollOptionVoteListener optionVoteListener;

    PollOptionAdapter(@NonNull PollOptionVoteListener optionVoteListener) {
        // A poll must have at least 2 options to choose between
        super(2);

        this.optionVoteListener = optionVoteListener;
    }

    void setPollVotingEnabled(boolean pollVotingEnabled) {
        this.pollVotingEnabled = pollVotingEnabled;
    }

    @Override
    public void onClick(PollOption item, int position) {
        if (pollVotingEnabled && !item.hasVoted) {
            optionVoteListener.onPollOptionClick(position);

            getItems().get(position).hasVoted = true;
            notifyItemChanged(position);
        }
    }

    @Override
    public PollOptionViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).
                inflate(R.layout.poll_vote_option_list_item, viewGroup, false);
        return new PollOptionViewHolder(itemView, this);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/poll/PollOptionAdapter.kt
package com.eulersbridge.isegoria.poll

import android.view.LayoutInflater
import android.view.ViewGroup
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.model.PollOption
import com.eulersbridge.isegoria.util.ui.LoadingAdapter

// A poll must have at least 2 options to choose between

internal class PollOptionAdapter internal constructor(private val optionVoteListener: PollOptionVoteListener)
    : LoadingAdapter<PollOption, PollOptionViewHolder>(2),
        PollOptionViewHolder.ClickListener {

    internal var pollVotingEnabled = true

    interface PollOptionVoteListener {
        /**
         * Called when any view of a poll option is clicked, but not when poll voting is disabled
         */
        fun onPollOptionClick(position: Int)
    }

    override fun onClick(item: PollOption, position: Int) {
        if (pollVotingEnabled && !item.hasVoted) {
            optionVoteListener.onPollOptionClick(position)

            items[position].hasVoted = true
            notifyItemChanged(position)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PollOptionViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.poll_vote_option_list_item,
            parent, false)
        return PollOptionViewHolder(itemView, this)
    }
}